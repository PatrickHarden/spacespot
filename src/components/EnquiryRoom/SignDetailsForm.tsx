import React, { useState, useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { useIntl } from 'react-intl'
import styled from 'styled-components'
import Input from 'components/common/Input'
import Heading2 from 'components/common/Heading2'
import Heading4 from 'components/common/Heading4'
import FilledButton from 'components/common/FilledButton'
import OutlinedButton from 'components/common/OutlinedButton'

import actions from 'services/negotiation/actions'
import { NegotiationStatus } from 'services/negotiation/types'
import { invalidString, invalidEmail } from 'services/global/util'
import { useSignerInfo } from 'services/negotiation/hooks'

const SignButtons = styled(
  (props: { className?: string; onNext: () => void; onBack: () => void }) => {
    const { className, onNext, onBack } = props
    const { formatMessage } = useIntl()
    const t = (s: string) => formatMessage({ id: s })
    return (
      <div className={className}>
        <FilledButton onClick={onNext}>
          {t('SUMMARY_CTA_CONTINUE')}
        </FilledButton>
        <OutlinedButton onClick={onBack}>
          {t('SUMMARY_CTA_CANCEL')}
        </OutlinedButton>
      </div>
    )
  },
)`
  display: flex;
  margin-top: 20px;
  border-top: 1px solid #dddddd;
  flex-direction: column;
  align-items: center;
  padding: 1em 0px;
  ${FilledButton} {
    width: 100%;
    background: #404042;
    margin-bottom: 5px;
  }
  ${OutlinedButton} {
    width: 100%;
    border: none;
    color: #404042;
    text-transform: none;
  }
`

const Subtitle = styled(Heading4)`
  margin-top: 20px;
  margin-bottom: 0;
`

const Desc = styled.p`
  margin: 0;
  color: #6d6c6c;
`

const FormBlock = styled.div`
  margin-top: 20px;
  margin-bottom: 40px;
`

const Row = styled.div`
  margin-right: -10px;
  display: flex;
  & > div {
    margin-right: 10px;
  }
  margin-bottom: 30px;
`

const SignDetailsForm = styled((props: { className?: string }) => {
  const dispatch = useDispatch()
  const { formatMessage } = useIntl()
  const t = (s: string) => formatMessage({ id: s })
  const { className } = props
  const signerInfo = useSignerInfo()
  const details0 = {
    name: '',
    emailId: '',
    companyNumber: '',
    companyName: '',
  }
  const [details, setDetails] = useState(details0)
  useEffect(() => {
    if (signerInfo) {
      setDetails(signerInfo)
    }
  }, [signerInfo, setDetails])

  const onChange = (name: string) => (
    e: React.ChangeEvent<HTMLInputElement>,
  ) => {
    setDetails({ ...details, [name]: e.target.value })
  }

  const [touched, setTouched] = useState(false)
  const validate = () => {
    const errors: { [key: string]: string } = {}
    if (invalidString(details.name))
      errors.name = t('SUMMARY_SIGNING_FORM_INVALID')
    if (invalidEmail(details.emailId))
      errors.emailId = t('SUMMARY_SIGNING_FORM_INVALID')
    if (invalidString(details.companyNumber))
      errors.companyNumber = t('SUMMARY_SIGNING_FORM_INVALID')
    if (invalidString(details.companyName))
      errors.companyName = t('SUMMARY_SIGNING_FORM_INVALID')
    return errors
  }

  const next = () => {
    setTouched(true)
    const errors = validate()
    const n = Object.keys(errors).length
    if (n === 0) {
      if (signerInfo) {
        dispatch(actions.putSignerInfo(details))
      } else {
        dispatch(actions.postSignerInfo(details))
      }
    }
  }

  const back = () => {
    dispatch(actions.putNegotiationStatus(NegotiationStatus.PreviewStart))
  }

  const errors: { [key: string]: string } = touched ? validate() : {}
  return (
    <div className={className}>
      <Heading2>{t('SUMMARY_SIGNING_FORM_TITLE')}</Heading2>
      <FormBlock>
        <Subtitle>{t('SUMMARY_SIGNING_FORM_PERSON')}</Subtitle>
        <Desc>{t('SUMMARY_SIGNING_FORM_PERSON_DSC')}</Desc>
        <Input
          id="name"
          onChange={onChange('name')}
          data-testid="name"
          type="text"
          label={t('SUMMARY_SIGNING_FORM_NAME')}
          value={details.name}
          min="0"
          error={errors.name}
        />
        <Input
          id="emailId"
          onChange={onChange('emailId')}
          data-testid="emailId"
          type="text"
          label={t('SUMMARY_SIGNING_FORM_EMAIL')}
          value={details.emailId}
          min="0"
          error={errors.emailId}
        />
      </FormBlock>
      <FormBlock>
        <Subtitle>{t('SUMMARY_SIGNING_FORM_LEGAL')}</Subtitle>
        <Desc>{t('SUMMARY_SIGNING_FORM_LEGAL_DSC')}</Desc>
        <Row>
          <Input
            id="companyName"
            onChange={onChange('companyName')}
            data-testid="companyName"
            type="text"
            label={t('SUMMARY_SIGNING_FORM_LEGAL_NAME')}
            value={details.companyName}
            min="0"
            error={errors.companyName}
          />
          <Input
            id="companyNumber"
            onChange={onChange('companyNumber')}
            data-testid="companyNumber"
            type="text"
            label={t('SUMMARY_SIGNING_FORM_LEGAL_NUMBER')}
            value={details.companyNumber}
            min="0"
            error={errors.companyNumber}
          />
        </Row>
      </FormBlock>
      <SignButtons onNext={next} onBack={back} />
    </div>
  )
})`
  margin: 0 20px;
`

export default SignDetailsForm
