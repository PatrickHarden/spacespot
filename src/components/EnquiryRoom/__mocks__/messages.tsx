export const messages = [
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'message',
    sentAt: '2018-10-18T12:40:35.886+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'message1',
    sentAt: '2018-10-18T12:40:43.547+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'message2',
    sentAt: '2018-10-18T12:40:44.946+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'message3',
    sentAt: '2018-10-18T12:40:45.678+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'message4',
    sentAt: '2018-10-18T12:40:46.856+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'message5',
    sentAt: '2018-10-23T14:17:59.662+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'jk',
    sentAt: '2018-10-28T10:13:59.805+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'message6',
    sentAt: '2018-10-28T10:16:16.585+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: '11:44',
    sentAt: '2018-10-28T10:45:02.592+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: '12:34',
    sentAt: '2018-10-28T11:23:27.467+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'll',
    sentAt: '2018-10-28T11:29:37.057+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: '12:30',
    sentAt: '2018-10-28T11:30:13.643+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'The message',
    sentAt: '2018-10-29T08:51:21.007+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'Another',
    sentAt: '2018-10-29T09:28:49.589+0000',
  },
  {
    userId: 'f9251bff-4527-4ebd-9028-f77b4bd0e0ce',
    message: 'Other',
    sentAt: '2018-10-29T09:29:02.900+0000',
  },
]
