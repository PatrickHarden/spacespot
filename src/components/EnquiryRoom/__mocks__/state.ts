export const mockState: any = {
  enquiry: {
    data: {
      spaceId: 'SS_SPACE_665',
      peer1Id: '1',
      peer2Id: '2',
      chatId: 159,
      enquireStatus: 'ACCEPTED',
      negotiationId: 139,
      enquireId: 159,
      flex: false,
    },
  },
  negotiation: {
    data: {
      flexible: false,
      terms: [
        {
          termId: 5,
          value: 'MONTHLY',
        },
        {
          termId: 6,
          value: '2000',
        },
        {
          termId: 7,
          value: '2',
        },
        {
          termId: 8,
          value: '2000',
        },
        {
          termId: 9,
          value: 'GUARANTEE',
        },
        {
          termId: 1,
          value: '3500',
        },
        {
          termId: 2,
          value: '40000',
        },
        {
          termId: 3,
          value: '2020-04-22',
        },
        {
          termId: 4,
          value: '13',
        },
      ],
      negotiationId: 139,
      enquireId: 159,
      createdAt: '2020-02-14T06:30:08.354+0000',
      lastUpdated: '2020-03-05T16:35:02.513+0000',
      lastModifyBy: '0062fb00-e389-4bb8-9119-613e3c45938a',
      status: 'PROPOSED',
      customFitout: false,
      fitoutOption: 0,
      specialProvision: 'gfd',
      specialProvisionStatus: 'CREATED',
    },
    loading: false,
    termsFields: {
      '1': 'Area',
      '2': 'Rent',
      '3': 'Start',
      '4': 'Duration',
      '5': 'Frequency',
      '6': 'AdditionalCosts',
      '7': 'Deposit',
      '8': 'ServiceCharges',
      '9': 'DepositType',
    },
    signerInfo: {
      id: 1,
      name: 'Name1',
      emailId: 'name1@email.com',
      companyName: 'companyA',
      companyNumber: '12345678',
    },
    customFitout: [],
  },
  user: {
    auth: {
      idToken: {
        claims: {
          // eslint-disable-next-line @typescript-eslint/camelcase
          extension_Landlord: true,
        },
      },
    },
  },
}
