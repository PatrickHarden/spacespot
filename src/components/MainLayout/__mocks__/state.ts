export const mockState: any = {
  user: {
    auth: {
      idToken: {
        claims: {
          // eslint-disable-next-line @typescript-eslint/camelcase
          extension_Landlord: true,
        },
      },
    },
  },
}
