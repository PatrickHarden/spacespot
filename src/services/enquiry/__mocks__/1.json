{
  "chatId": 1,
  "enquireId": 1,
  "spaceId": 2,
  "peer1Id": "8fa47f17-c44e-48eb-8e88-9e8c8fe5cda3",
  "peer2Id": "8fa47f17-c44e-48eb-8e88-9e8c8fe5cda4"
}
