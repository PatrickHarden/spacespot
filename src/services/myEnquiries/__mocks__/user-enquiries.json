[
  {
    "enquireId": 167,
    "negotiationId": 147,
    "enquireStatus": "PENDING_TO_ACCEPT",
    "spaceId": "SS_SPACE_650",
    "terms": [
      { "termId": 1, "value": "12" },
      { "termId": 2, "value": "12.0" },
      {
        "termId": 3,
        "value": "Tue Feb 25 2020 17:59:47 GMT+0100 (hora estándar de Europa central)"
      },
      { "termId": 4, "value": "123" },
      { "termId": 5, "value": "" },
      { "termId": 6, "value": "" },
      { "termId": 7, "value": "" },
      { "termId": 8, "value": "123.0" },
      { "termId": 9, "value": "" }
    ],
    "messages": [
      {
        "userId": "de2563f1-2e15-4a6a-9e39-9600a8ad4ec1",
        "message": "",
        "sentAt": "2020-02-25T16:59:53.122+0000"
      }
    ]
  },
  {
    "enquireId": 168,
    "negotiationId": 148,
    "enquireStatus": "PENDING_TO_ACCEPT",
    "spaceId": "SS_SPACE_650",
    "terms": [
      { "termId": 1, "value": "12" },
      { "termId": 2, "value": "12.0" },
      {
        "termId": 3,
        "value": "Tue Feb 25 2020 18:00:28 GMT+0100 (hora estándar de Europa central)"
      },
      { "termId": 4, "value": "13" },
      { "termId": 5, "value": "" },
      { "termId": 6, "value": "" },
      { "termId": 7, "value": "" },
      { "termId": 8, "value": "123.0" },
      { "termId": 9, "value": "" }
    ],
    "messages": [
      {
        "userId": "de2563f1-2e15-4a6a-9e39-9600a8ad4ec1",
        "message": "",
        "sentAt": "2020-02-25T17:00:34.014+0000"
      }
    ]
  }
]
