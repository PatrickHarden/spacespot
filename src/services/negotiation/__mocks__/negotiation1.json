{
  "terms": [
    { "termId": 1, "value": "34" },
    { "termId": 2, "value": "207" },
    { "termId": 3, "value": "29 Months" },
    { "termId": 4, "value": "200 Sq Mtr" },
    { "termId": 5, "value": "$5600 per month" },
    { "termId": 6, "value": "20th September" }
  ],
  "negotiationId": 129,
  "enquireId": 149,
  "createdAt": "2020-02-03T13:03:07.661+0000",
  "lastUpdated": "2020-02-11T09:06:15.764+0000",
  "lastModifyBy": "c700793a-1963-4b2b-996f-1939f853cc10",
  "status": "CREATED"
}
