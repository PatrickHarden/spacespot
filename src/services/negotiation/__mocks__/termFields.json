{
  "1": "Area",
  "2": "Rent",
  "3": "Start",
  "4": "Duration",
  "5": "Frequency",
  "6": "AdditionalCosts",
  "7": "Deposit",
  "8": "ServiceCharges"
}
