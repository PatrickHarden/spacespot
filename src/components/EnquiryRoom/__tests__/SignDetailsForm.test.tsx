import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { createMockStore } from 'redux-test-utils'
import { Provider } from 'react-redux'

import SignDetailsForm from '../SignDetailsForm'

import { mockState } from '../__mocks__/state'

jest.mock('react-intl', () => ({
  useIntl: () => ({ formatMessage: (s: any) => s.id }),
  createIntlCache: () => jest.fn(),
  createIntl: () => ({ formatMessage: (s: any) => s.id }),
}))

const mockDispatch = jest.fn()

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}))

it('Renders without crash', () => {
  const store: any = createMockStore(mockState)
  const wrapper = render(
    <Provider store={store}>
      <SignDetailsForm />
    </Provider>,
  )
  expect(wrapper.asFragment()).toMatchSnapshot()
  fireEvent.change(wrapper.getByTestId('name'), { target: { value: '' } })
  fireEvent.change(wrapper.getByTestId('emailId'), { target: { value: '' } })
  fireEvent.change(wrapper.getByTestId('companyName'), {
    target: { value: '' },
  })
  fireEvent.change(wrapper.getByTestId('companyNumber'), {
    target: { value: '' },
  })
  fireEvent.click(wrapper.getByText('SUMMARY_CTA_CONTINUE'))
  fireEvent.change(wrapper.getByTestId('name'), { target: { value: 'Name1' } })
  fireEvent.change(wrapper.getByTestId('emailId'), {
    target: { value: 'test@email.com' },
  })
  fireEvent.change(wrapper.getByTestId('companyName'), {
    target: { value: 'Company1' },
  })
  fireEvent.change(wrapper.getByTestId('companyNumber'), {
    target: { value: '1234567' },
  })
  fireEvent.click(wrapper.getByText('SUMMARY_CTA_CONTINUE'))
  fireEvent.click(wrapper.getByText('SUMMARY_CTA_CANCEL'))
})

it('Posts signerInfo', () => {
  mockState.negotiation.signerInfo = null
  const store: any = createMockStore(mockState)

  const w = render(
    <Provider store={store}>
      <SignDetailsForm />
    </Provider>,
  )
  fireEvent.change(w.getByTestId('name'), { target: { value: 'Name1' } })
  fireEvent.change(w.getByTestId('emailId'), {
    target: { value: 'test@email.com' },
  })
  fireEvent.change(w.getByTestId('companyName'), {
    target: { value: 'Company1' },
  })
  fireEvent.change(w.getByTestId('companyNumber'), {
    target: { value: '1234567' },
  })
  fireEvent.click(w.getByText('SUMMARY_CTA_CONTINUE'))
})
