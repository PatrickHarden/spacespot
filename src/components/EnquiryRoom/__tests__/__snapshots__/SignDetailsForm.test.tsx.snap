// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without crash 1`] = `
<DocumentFragment>
  <div
    class="sc-gqjmRU hurzXi"
  >
    <h2
      class="sc-bxivhb fTAvxB"
    >
      SUMMARY_SIGNING_FORM_TITLE
    </h2>
    <div
      class="sc-iwsKbI jilomH"
    >
      <h4
        class="sc-ifAKCX sc-htoDjs cudOw"
      >
        SUMMARY_SIGNING_FORM_PERSON
      </h4>
      <p
        class="sc-dnqmqq epmrXp"
      >
        SUMMARY_SIGNING_FORM_PERSON_DSC
      </p>
      <div
        class="sc-bdVaJa dlaMlh"
      >
        <label
          for="name"
        >
          SUMMARY_SIGNING_FORM_NAME
        </label>
        <input
          class="sc-bwzfXH jHsaxX"
          data-testid="name"
          id="name"
          min="0"
          type="text"
          value="Name1"
        />
        <div
          class="sc-htpNat gctktF"
        />
      </div>
      <div
        class="sc-bdVaJa dlaMlh"
      >
        <label
          for="emailId"
        >
          SUMMARY_SIGNING_FORM_EMAIL
        </label>
        <input
          class="sc-bwzfXH jHsaxX"
          data-testid="emailId"
          id="emailId"
          min="0"
          type="text"
          value="name1@email.com"
        />
        <div
          class="sc-htpNat gctktF"
        />
      </div>
    </div>
    <div
      class="sc-iwsKbI jilomH"
    >
      <h4
        class="sc-ifAKCX sc-htoDjs cudOw"
      >
        SUMMARY_SIGNING_FORM_LEGAL
      </h4>
      <p
        class="sc-dnqmqq epmrXp"
      >
        SUMMARY_SIGNING_FORM_LEGAL_DSC
      </p>
      <div
        class="sc-gZMcBi QWWJs"
      >
        <div
          class="sc-bdVaJa dlaMlh"
        >
          <label
            for="companyName"
          >
            SUMMARY_SIGNING_FORM_LEGAL_NAME
          </label>
          <input
            class="sc-bwzfXH jHsaxX"
            data-testid="companyName"
            id="companyName"
            min="0"
            type="text"
            value="companyA"
          />
          <div
            class="sc-htpNat gctktF"
          />
        </div>
        <div
          class="sc-bdVaJa dlaMlh"
        >
          <label
            for="companyNumber"
          >
            SUMMARY_SIGNING_FORM_LEGAL_NUMBER
          </label>
          <input
            class="sc-bwzfXH jHsaxX"
            data-testid="companyNumber"
            id="companyNumber"
            min="0"
            type="text"
            value="12345678"
          />
          <div
            class="sc-htpNat gctktF"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-gzVnrw IKReL"
    >
      <button
        class="sc-EHOje hOldrS"
      >
        SUMMARY_CTA_CONTINUE
      </button>
      <button
        class="sc-bZQynM gAEiOa"
      >
        SUMMARY_CTA_CANCEL
      </button>
    </div>
  </div>
</DocumentFragment>
`;
